Classes/GlobalStateExplorers/DatabaseBrowser/FLEXDatabaseManager.m

```diff
@@ -0,0 +1,194 @@
+//
+//  PTDatabaseManager.m
+//  PTDatabaseReader
+//
+//  Created by Peng Tao on 15/11/23.
+//  Copyright © 2015年 Peng Tao. All rights reserved.
+//
+
+#import "FLEXDatabaseManager.h"
+#import <sqlite3.h>
+
+
+static NSString *const QUERY_TABLENAMES_SQL = @"SELECT name FROM sqlite_master WHERE type='table' ORDER BY name";
+
+
+@implementation FLEXDatabaseManager
+{
+  sqlite3* _db;
+  NSString* _databasePath;
+}
+
+
+- (instancetype)initWithPath:(NSString*)aPath
+{  
+  self = [super init];
+  
+  if (self) {
+    _databasePath = [aPath copy];
+    _db           = nil;
+  }
+  return self;
+}
+
+
+- (BOOL)open {
+  if (_db) {
+    return YES;
+  }
+  int err = sqlite3_open([_databasePath UTF8String], &_db);
+  if(err != SQLITE_OK) {
+    NSLog(@"error opening!: %d", err);
+    return NO;
+  }
+  return YES;
+}
+
+- (BOOL)close {
+  if (!_db) {
+    return YES;
+  }
+  
+  int  rc;
+  BOOL retry;
+  BOOL triedFinalizingOpenStatements = NO;
+  
+  do {
+    retry   = NO;
+    rc      = sqlite3_close(_db);
+    if (SQLITE_BUSY == rc || SQLITE_LOCKED == rc) {
+      if (!triedFinalizingOpenStatements) {
+        triedFinalizingOpenStatements = YES;
+        sqlite3_stmt *pStmt;
+        while ((pStmt = sqlite3_next_stmt(_db, nil)) !=0) {
+          NSLog(@"Closing leaked statement");
+          sqlite3_finalize(pStmt);
+          retry = YES;
+        }
+      }
+    }
+    else if (SQLITE_OK != rc) {
+      NSLog(@"error closing!: %d", rc);
+    }
+  }
+  while (retry);
+  
+  _db = nil;
+  return YES;
+}
+
+
+- (NSArray *)queryAllTables
+{
+  return [self executeQuery:QUERY_TABLENAMES_SQL];
+}
+
+- (NSArray *)queryAllColumnsWithTableName:(NSString *)tableName
+{
+  NSString *sql = [NSString stringWithFormat:@"PRAGMA table_info('%@')",tableName];
+  NSArray *resultArray =  [self executeQuery:sql];
+  NSMutableArray *array = [NSMutableArray array];
+  for (NSDictionary *dict in resultArray) {
+    [array addObject:dict[@"name"]];
+  }
+  return array;
+}
+
+- (NSArray *)queryAllDataWithTableName:(NSString *)tableName
+{
+  NSString *sql = [NSString stringWithFormat:@"SELECT * FROM %@",tableName];
+  return [self executeQuery:sql];
+}
+
+#pragma mark -
+#pragma mark - Private
+
+- (NSArray *)executeQuery:(NSString *)sql
+{
+  [self open];
+  NSMutableArray *resultArray = [NSMutableArray array];
+  sqlite3_stmt *pstmt;
+  if (sqlite3_prepare_v2(_db, [sql UTF8String], -1, &pstmt, 0) == SQLITE_OK) {
+    while (sqlite3_step(pstmt) == SQLITE_ROW) {
+      NSUInteger num_cols = (NSUInteger)sqlite3_data_count(pstmt);
+      if (num_cols > 0) {
```

Comment:
`num_cols` will always be `> 0` because it is unsigned. Maybe make it an `int` since that's what `sqlite3_data_count()` returns. Small nit on style too: prefer camel case instead of underscores for variable names.